Provide strict compliance for DynaLoader and XSLoader

toddr opened the pull request:
Add warnings to tests.

toddr committed:
Provide strict compliance for DynaLoader and XSLoader

Add warnings to tests.

toddr commented on ext/DynaLoader/DynaLoader_pm.PL:
```diff
@@ -292,12 +298,14 @@ sub croak   { require Carp; Carp::croak(@_)   }
-    local @isa = (@isa, 'DynaLoader');
+
+    no strict 'refs';
+    push @{"$module\::ISA"}, 'DynaLoader';
```
I consider this line to be the riskiest part of this change. I'd love someone to confirm the change is ok.

Grinnz replied:
This changes that module's `@ISA` permanently, rather than only while running `bootstrap`. There's also the chance that something is relying on `@Dynaloader::isa` being set while running `bootstrap`, this seems unlikely but I'm not that familiar with how this module works. The original code should be fine with `no strict 'refs'` and declaring `our @isa`.

hvds replied:
Agree on all counts with @Grinnz, he's too quick.

Leont replied:
Yeah, the `local` here is absolutely essential. We only want the modules to inherit from `DynaLoader` during bootstrap, so that it has a default `dl_load_flags` implementation during that bootstrap. This design is unfortunate, but no one really knew how to OO back in 1994.

toddr replied:
ok. I put it back.

haarg commented on dist/XSLoader/XSLoader_pm.PL:
```diff
@@ -9,11 +11,13 @@ print OUT <<'EOT';
-package XSLoader;
+use strict;
+no strict 'refs';
+#use warnings;
```
These warnings comments probably either need some explanation or should be omitted.

toddr replied:
It's meant for a follow on discussion. I'll pull it out for now if it's distracting 😄

haarg reviewed (changes requested):
The `print OUT <<'EOT' if $] < 5.006;` code block won't work with strict on.

Grinnz commented on ext/DynaLoader/DynaLoader_pm.PL:
```diff
@@ -292,12 +298,14 @@ sub croak   { require Carp; Carp::croak(@_)   }
+
+    no strict 'refs';
+    push @{"$module\::ISA"}, 'DynaLoader';
     # Cannot goto due to delocalization.  Will report errors on a wrong line?
```
For the record I believe this comment's second concern is unwarranted, Carp does not consider calls from within the same package when reporting errors.

toddr replied:
This seems out of scope for this work. I'm happy to take a follow on commit or help with a different PR.

Leont commented on ext/DynaLoader/DynaLoader_pm.PL:
```diff
@@ -292,12 +298,14 @@ sub croak   { require Carp; Carp::croak(@_)   }
     bootstrap(@_);
 }
 
+our (@args, $module, @dirs, $file);
```
This declaration probably makes more sense inside sub bootstrap?

toddr replied:
I couldn't decide if any of these variables were needed so I left them in the stash. Happy to move them into the sub.

toddr replied:
If I put them in sub bootstrap, then they're still `@DynaLoader::args`, right? Are we just discussing style or a functional difference here? Note the local statements following this line.

Grinnz replied:
The lexical alias would only exist in the scope of the sub. It would alias to the same package variables either way.

toddr replied:
ok.

toddr commented:
> The `print OUT <<'EOT' if $] < 5.006;` code block won't work with strict on.

Can you show me how this is different with and without strict?
```
$>perl -e 'print STDERR <<'EOT' if $] > 5.006        
foo
foo
EOT'
foo
foo

$>perl -e 'use strict; print STDERR <<'EOT' if $] > 5.006
foo
foo
EOT'
foo
foo
```

toddr commented on ext/DynaLoader/DynaLoader_pm.PL:
```diff
@@ -550,6 +558,8 @@ sub dl_find_symbol_anywhere
     return undef;
 }
 
+1;
```
I've spent a bunch of time questioning my understanding of how to make a module true. 

I think this line (much higher up) is making the module true and I get it but it still surprises me that it works. 

```
1; # End of main code
```

toddr replied:
This of course blames to Larry showing off in a0d0e21ea6e

Grinnz replied:
Specifically it's the file that needs to return true when required. The sub definitions are evaluated at parse time so don't affect the "return value".

toddr commented:
> The `print OUT <<'EOT' if $] < 5.006;` code block won't work with strict on.

What you're actually saying is that the generated code for Perl below 5.6 won't be strict?

This code won't run below 5.6 because of our so that's probably moot. I guess it should be removed.

toddr committed:
Drop support for XSLoader below 5.6

Now we're using our, there's no value to attempting to support
versions below 5.6.

toddr commented:
> This code won't run below 5.6 because of our so that's probably moot. I guess it should be removed. Even looking at testers, the lowest report is 5.6 and there's only 1.

I've added a second commit to remove code supporting 5.6

ilmari commented on ext/DynaLoader/DynaLoader_pm.PL:
```diff
@@ -300,6 +307,7 @@ sub bootstrap_inherit {
 
 sub bootstrap {
     # use local vars to enable $module.bs script to edit values
+    our (@args, $module, @dirs, $file);
     local(@args) = @_;
     local($module) = $args[0];
     local(@dirs, $file);
```
I think it would be neater to add the `our` to the `local` statements, to avoid repeating the variable names. e.g.:
```perl
    local our @args = @_;
    local our $module = $args[0];
    local our (@dirs, $file);
```

toddr replied:
Done.

ilmari commented on dist/XSLoader/Makefile.PL:
```diff
@@ -8,6 +8,8 @@
 use strict;
 use warnings;
 
+require 5.006;
```
Why not `use v5.6;`?

Grinnz replied:
There's not much point in using the vstring form of a version for a 5.6 requirement since 5.6 is the first perl version that recognizes it.

ilmari replied:
Point. But there's no reason to use `require` instead of `use`.

toddr replied:
IMO, A better bike shed is:

> what should dual life modules be doing to specify minimum version?

I would argue we're REALLY inconsistent with the use of these lines. In all honesty, I'm not convinced they have much value. It'd be good to have a documented standard for how we want to use this in our code. After which we can just make one big change to fix them all.

Grinnz replied:
This is not a bike shed, there is a functional difference between the suggestions.

haarg replied:
`require` is most likely a remnant of history, since 5.003 supported a perl version check using `require` but not `use`.

`use v5.6;` or `use 5.006;` both seem reasonable to me. `v5.6` is more human readable. The only real downside I see is that version conditionals are best done in the form `"$]" >= 5.008001`, so they wouldn't be consistent. That's very minor though.

toddr replied:
done.

Leont commented:
On second thought, I think this whole approach is probably wrong for the most unfortunate and unlikely of reasons: we need use «use vars»

I know what you're thinking, _what drugs is he on?_, but please hear me out. Many of these variables are supposed to be available for the bootstrap files, and non-lexically enabling that is exactly what use vars does (as explained [here](https://rjbs.manxome.org/rubric/entry/1862) by rjbs).

toddr commented:
> Otherwise `.bs` files and `_BS` files are likely to not compile when strict is enabled.

I've done that count. There are literally 4 bs non-zero sized files on CPAN. Those seem fixable. Don't get me wrong. I'm not trying to dismiss this problem. But I think it's fixable if that's the only issue.

```
-rw-r--r-- 1    100 games  326 Dec  1  1996 M/Meschach/Meschach.bs
-rw-r--r-- 1 centos centos 182 Sep 15 02:43 D/DB_File/DB_File_BS
-rw-r--r-- 1    500 root   306 Sep 10  2012 E/Embperl/Embperl_BS
-rw-r--r-- 1 root   root   304 Dec  4  2001 H/HTML-Embperl/Embperl_BS
```

https://st.aticpan.org/source/EGROSS/Meschach-0.03/Meschach.bs
https://github.com/Perl/perl5/blob/blead/cpan/DB_File/DB_File_BS
https://st.aticpan.org/source/GRICHTER/Embperl-2.5.0/Embperl_BS
https://st.aticpan.org/source/GRICHTER/HTML-Embperl-1.3.6/Embperl_BS

toddr commented:
Since DB_File is a dual life modulel, I've opened a case to remove the BS file. https://github.com/pmqs/DB_File/pull/7

pmqs commented:
> Since DB_File is a dual life modulel, I've opened a case to remove the BS file. [pmqs/DB_File#7](https://github.com/pmqs/DB_File/pull/7)

Change merged upstream

toddr commented:
@Leont I've been looking closer at what Embperl is doing. I don't see anything there that wouldn't be better placed in a BEGIN block that makes a decision to die or just not load the module. Do you agree?

jkeenan commented:
DynaLoader:  Setting aside one merge conflict (a `$VERSION` change), where do we stand with this pull request?  Should we expect to merge it in this dev cycle?

@LeonT, @toddr, can you provide status update?

Thank you very much.
Jim Keenan

toddr commented:
@jkeenan thanks for the ping. I've rebased this branch but not really looked closely to make sure there are no further issues. Looks like I was waiting for leon to respond and never heard back from him. I can look at this branch next week when work allows. 

Thanks!

toddr commented:
Status update:
Leon's concerns seemed to be .bs files which are pretty much dead and should probably be deprecated for non-use. 

Only 4 examples of non-zero .bs files exist on CPAN:
* DB_File which no longer uses .bs files as of a recent release.
* Meschach hasn't been updated since 2003
* Embperl hasn't passed tests (or even compiled?) since 5.18.
* HTML::Embperl only passes tests on 5.6 and hasn't had a release since 2003. 

So I repeat my question to @Leont: Is this something we're really concerned about?

Leont commented:
> So I repeat my question to @Leont: Is this something we're really concerned about?

It's not that I'm terribly concerned, but I already pointed out how we can make this strict without breaking any usage of this.

jkeenan commented:
@Leont's p.r. in https://github.com/Perl/perl5/pull/18662 was approved by @tonycoz 6 days ago.  I rebased and retested it before merging it to blead just now.  That (I believe) obviates the need for this p.r., so I'm closing it.  Any remaining problems we can deal with in a new issue or pull request.

Thank you very much.
Jim Keenan
